doc/cascadia/Keybindings-Arguments.md

```diff
@@ -0,0 +1,315 @@
+// etc...
+```
+
+Note that instead of having 9 different `newTabProfile<N>` actions, we have a
+singular `newTabProfile` action, and that action requires a `profileIndex` in
+the `args` object.
+
+Also, pay attention to the last set of keybindings, the `splitProfile` ones.
+This is a function that requires two arguments, both a `orientation` and a
+`profileIndex`. Before this change we would have needed to create 20 separate
+actions (10 profile indicies * 2 directions) to handle these cases. Now it can
+be done with a single action that can be much more flexible in its
+implementation.
+
+### Parsing KeyBinding Arguments
+
+We'll add a new `IKeyBindingArgs` interface. All current keybinding events will
+be changed from their current types to `TypedEventHandler`s. These
+`TypedEventHandler`s second param will always be an instance of
+`IKeyBindingArgs`. So for example:
+
+```csharp
+
+delegate void CopyTextEventArgs();
+delegate void NewTabEventArgs();
+delegate void NewTabWithProfileEventArgs(Int32 profileIndex);
+// ...
+
+[default_interface]
+runtimeclass AppKeyBindings : Microsoft.Terminal.Settings.IKeyBindings
+{
+    event CopyTextEventArgs CopyText;
+    event NewTabEventArgs NewTab;
+    event NewTabWithProfileEventArgs NewTabWithProfile;
+```
+
+Becomes:
+
+```csharp
+interface IKeyBindingArgs {
+    Boolean Handled;
+}
+
+runtimeclass CopyTextEventArgs : IKeyBindingArgs
+{
+    Boolean CopyWhitespace;
+}
+runtimeclass NewTabEventArgs : IKeyBindingArgs { }
+runtimeclass NewTabWithProfileEventArgs : IKeyBindingArgs
+{
+    Int32 ProfileIndex;
+}
+
+[default_interface]
+runtimeclass AppKeyBindings : Microsoft.Terminal.Settings.IKeyBindings
+{
+    event Windows.Foundation.TypedEventHandler<AppKeyBindings, CopyTextEventArgs> CopyText;
+    event Windows.Foundation.TypedEventHandler<AppKeyBindings, NewTabEventArgs> NewTab;
+    event Windows.Foundation.TypedEventHandler<AppKeyBindings, NewTabWithProfileEventArgs> NewTabWithProfile;
+```
+
+We'll also change our existing map in the `AppKeyBindings` implementation.
+Currently, it's a `std::unordered_map<KeyChord, ShortcutAction, ...>`, which
+uses the `KeyChord` to lookup the `ShortcutAction`. We'll need to introduce a
+new type `ActionAndArgs`:
+
+```csharp
+runtimeclass ActionAndArgs
+{
+    ShortcutAction action;
+    IKeyBindingArgs args;
+}
+```
+
+and we'll change the map in `AppKeyBindings` to a `std::unordered_map<KeyChord,
+ActionAndArgs, ...>`.
+
+When we're parsing keybindings, we'll need to construct args for each of the
+events to go with each binding. When we find some key chord bound to a given
+Action, we'll construct the `IKeyBindingArgs` for that action. For many actions,
+these args will be an empty class. However, when we do find an action that needs
+additional parsing, `AppKeyBindingsSerialization` will do the extra work to
+parse the args for that action.

```

Comment:
This doesn't work with _Handled_ actually. You might need two levels:

TypedEventHandler: Sender, IKeyBindingEventArgs

IKeyBindingEventArgs: Handled, IKeyBindingArgs

IKeyBindingArgs: The actual thing that was deserialized. The receiver must QI for the right interface?

This also reduces the chance that we muck up the implementation of Handled on one of them, because it's _outside_ the "must-be-implemented-per-event-type" class.

Reply:
(It doesn't work for Handled because that's per _event_, not per _binding.)

Reply:
Yea that makes sense, I'll add another layer.